---
title: "H3K4me1 vs H3K27ac"
author: "Chenghao Zhu"
date: "1/6/2020"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, warning=FALSE, message=FALSE}
library(dplyr)
library(reshape2)
library(ggplot2)
library(rtracklayer)
library(ChIPseeker)
library(clusterProfiler)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(org.Hs.eg.db)
library(biomaRt)
library(ggplot2)
theme_set(theme_classic())
```

Read narrowPeaks

```{r}
h3k4me1_tx = import(
    gzfile("data/H3K4me1_SR8278.merged.nodup_x_ctl_for_rep1.pval0.01.500K.narrowPeak.gz"),
    format = "narrowPeak"
)
h3k4me1_tx = h3k4me1_tx[h3k4me1_tx$qValue > 2]
h3k27ac_tx = import(
    gzfile("data/H3K27ac_SR8278.merged.nodup_x_ctl_for_rep1.pval0.01.500K.narrowPeak.gz"),
    format = "narrowPeak"
)
h3k27ac_tx = h3k27ac_tx[h3k27ac_tx$qValue > 2]
h3k4me1_ve = import(
    gzfile("data/H3K4me1_Vehicle.merged.nodup_x_ctl_for_rep2.pval0.01.500K.narrowPeak.gz"),
    format = "narrowPeak"
)
h3k4me1_ve = h3k4me1_ve[h3k4me1_ve$qValue > 2]
h3k27ac_ve = import(
    gzfile("data/H3K27ac_Vehicle.merged.nodup_x_ctl_for_rep2.pval0.01.500K.narrowPeak.gz"),
    format = "narrowPeak"
)
h3k27ac_ve = h3k27ac_ve[h3k27ac_ve$qValue > 2]
```

## Find overlap peaks

```{r}
overlaps_tx = findOverlapPairs(h3k27ac_tx, h3k4me1_tx)
overlaps_tx
```

There are 12,963 overlapped peaks between h3k27ac and h3k4me1 after treatment. 

```{r}
overlaps_ve = findOverlapPairs(h3k27ac_ve, h3k27ac_ve)
overlaps_ve
```

There are 82,684 overlapped peaks between h3k27ac and h3k4me1 in control. 

```{r}
h3k27ac_tx_overlap = length(unique(overlaps_tx@first$name))
h3k4me1_tx_overlap = length(unique(overlaps_tx@second$name))
h3k27ac_ve_overlap = length(unique(overlaps_ve@first$name))
h3k4me1_ve_overlap = length(unique(overlaps_ve@second$name))
data.frame(
    modification = c("h3k27ac", "h3k27ac", "h3k4me1", "h3k4me1", 
                "h3k27ac", "h3k27ac", "h3k4me1", "h3k4me1"),
    type = c("unique", "overlap", "unique", "overlap",
             "unique", "overlap", "unique", "overlap"),
    treatment = rep(c("tx", "ve"), each = 4),
    npeaks = c(length(h3k27ac_tx) - h3k27ac_tx_overlap, h3k27ac_tx_overlap, 
               length(h3k4me1_tx) - h3k4me1_tx_overlap, h3k4me1_tx_overlap,
               length(h3k27ac_ve) - h3k27ac_ve_overlap, h3k27ac_ve_overlap, 
               length(h3k4me1_ve) - h3k4me1_ve_overlap, h3k4me1_ve_overlap),
    anno.y = c((length(h3k27ac_tx) - h3k27ac_tx_overlap)/2 + h3k27ac_tx_overlap, 
               h3k27ac_tx_overlap/2, 
               (length(h3k4me1_tx) - h3k4me1_tx_overlap)/2 + h3k4me1_tx_overlap, 
               h3k4me1_tx_overlap/2,
               (length(h3k27ac_ve) - h3k27ac_ve_overlap)/2 + h3k27ac_ve_overlap, 
               h3k27ac_ve_overlap/2, 
               (length(h3k4me1_ve) - h3k4me1_ve_overlap)/2 + h3k4me1_ve_overlap, 
               h3k4me1_ve_overlap/2)
) %>%
    mutate(type = relevel(type, ref = "unique"),
           anno.y = anno.y + 100) %>%
    ggplot(aes(x = modification, y = npeaks, fill = type)) +
    geom_col(width = 0.5) +
    geom_text(aes(y = anno.y, label = npeaks)) +
    facet_grid(cols = vars(treatment))
```

For SR8278 treatment, h3k27ac has in total of 84,419 peaks, of which 11,571 overlaps with at least one peak from h3k4m21. h3k4me1 has in total of 93,660 peak, of which 12,353 overlaps with at least one peak from h3k27ac. As for vehicel control, h3k27ac has 82,684 peaks and all of them overlap with a peak from h3k4me1.

The total number of peaks in h3k27ac overlap with h3k4me1 does not match with that of h3k4me1 with h3k27ac, because some peaks overlaps with more than one peak from the other modification.

```{r}
df = merge(
    as.data.frame(table(overlaps_tx@first@seqnames@values)),
    as.data.frame(table(overlaps_ve@first@seqnames@values)),
    by = "Var1"
)
colnames(df) = c("chrom", "tx", "ve")
df %>% filter(chrom %in% paste0("chr", c(1:22, "X", "Y"))) %>% 
    arrange(ve) %>% 
    mutate(chrom = factor(chrom, levels = chrom)) %>%
    melt(id.vars = "chrom", value.name = "freq") %>%
    ggplot() + geom_col(aes(x = chrom, y = freq, fill = variable), position = "dodge") +
    coord_flip()
```

## Gene annotation

Obtain gene annotations

```{r}
chroms_ensembl = seqlevels(h3k27ac_tx)
chroms = gsub("^chr", "", chroms_ensembl)
chroms = gsub("^.+?_", "", chroms)
chroms = gsub("_.+$", "", chroms)
chroms = gsub("v([0-9])$", "\\.\\1", chroms)

ensembl = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
anno = getBM(
    mart = ensembl,
    filter = "chromosome_name",
    values = chroms,
    attributes = c("chromosome_name", "start_position", "end_position", 
                   "ensembl_gene_id", "external_gene_name", "gene_biotype")
)

anno_granges = GRanges(
    seqnames = chroms_ensembl[match(anno$chromosome_name, chroms)],
    ranges = IRanges(
        start = anno$start_position,
        end = anno$end_position,
        names = anno$ensembl_gene_id
    ),
    gene_name = anno$external_gene_name,
    gene_type = anno$gene_biotype
)
```

```{r}
tx_ranges = lapply(seq_along(overlaps_tx), function(i){
    start1 = start(overlaps_tx@first[i])
    end1 = end(overlaps_tx@first[i])
    start2 = start(overlaps_tx@second[i])
    end2 = end(overlaps_tx@second[i])
    res = c(min(start1, start2),
            max(end1, end2))
    names(res) = c("start", "end")
    return(res)
}) %>%
    do.call(rbind, .)
tx_granges = GRanges(seqnames = overlaps_tx@first@seqnames, 
                     ranges = IRanges(start = tx_ranges[,1], end = tx_ranges[,2]))
anno_tx_overlaps = findOverlaps(tx_granges, anno_granges)
anno_tx = anno_granges
anno_tx$n_overlap = as.integer(table(anno_tx_overlaps@to)[as.character(1:length(anno_tx))])
anno_tx = anno_tx[which(!is.na(anno_tx$n_overlap))]
anno_tx = anno_tx[order(anno_tx$n_overlap, decreasing = TRUE)]
anno_tx
```

```{r}
ve_ranges = lapply(seq_along(overlaps_ve), function(i){
    start1 = start(overlaps_ve@first[i])
    end1 = end(overlaps_ve@first[i])
    start2 = start(overlaps_ve@second[i])
    end2 = end(overlaps_ve@second[i])
    res = c(min(start1, start2),
            max(end1, end2))
    names(res) = c("start", "end")
    return(res)
}) %>%
    do.call(rbind, .)
ve_granges = GRanges(seqnames = overlaps_ve@first@seqnames, 
                     ranges = IRanges(start = ve_ranges[,1], end = ve_ranges[,2]))
anno_ve_overlaps = findOverlaps(ve_granges, anno_granges)
anno_ve = anno_granges
anno_ve$n_overlap = as.integer(table(anno_ve_overlaps@to)[as.character(1:length(anno_ve))])
anno_ve = anno_ve[which(!is.na(anno_ve$n_overlap))]
anno_ve = anno_ve[order(anno_ve$n_overlap, decreasing = TRUE)]
anno_ve
```

```{r}
gene_df = left_join(
    data.frame(gene_id   = names(anno_ve),
               gene_name = anno_ve$gene_name, 
               n_overlap = anno_ve$n_overlap,
               stringsAsFactors = FALSE),
    data.frame(gene_id   = names(anno_tx),
               gene_name = anno_tx$gene_name, 
               n_overlap = anno_tx$n_overlap,
               stringsAsFactors = FALSE), 
    by = c("gene_id", "gene_name"), 
)
gene_df$n_overlap.y[is.na(gene_df$n_overlap.y)] = 0
colnames(gene_df) = c("gene_id", "gene_name", "ve", "tx")
gene_df = mutate(gene_df, diff = ve - tx) %>%
    arrange(desc(diff))
gene_df 
```


```{r}
gene_df %>% head(20) %>%
    dplyr::select(-diff, -gene_id) %>%
    mutate(gene_name = factor(gene_name, levels = gene_name)) %>%
    melt(id.vars = "gene_name", variable.name = "treatment", value.name = "npeak") %>%
    ggplot() +
    geom_col(aes(gene_name, npeak, fill = treatment), position = "dodge") +
    coord_flip()
```

## GO enrichment

```{r}
library(clusterProfiler)
gene_down = bitr(gene_df$gene_id[gene_df$diff > 0], fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db, drop = TRUE)
gene_bg = bitr(gene_df$gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db, drop = TRUE)
ego = enrichGO(
    gene = gene_down$ENTREZID,
    OrgDb = org.Hs.eg.db,
    ont = "ALL",
    pvalueCutoff = 0.01,
    qvalueCutoff = 0.05,
    readable = TRUE
)
# write.table(ego@result, file = "go_enrichment.txt", sep = "\t", quote = FALSE, row.names = FALSE)
ego@result %>% dplyr::select(-ID, -geneID) %>% 
    DT::datatable() %>%
    DT::formatSignif(columns = 5:7, digits = 3)
```

Download the go enrichment table:

<a href="https://raw.githubusercontent.com/zhuchcn/sr8278/master/go_enrichment.txt" class="btn btn-primary">Download</a>

## Export data

```{r}
library(jsonlite)
names(gene_df)[3:4] = c("npeaks_vs", "npeaks_tx")
entrez = bitr(gene_df$gene_id, fromType = "ENSEMBL", toType = c("ENSEMBL", "ENTREZID"), OrgDb = org.Hs.eg.db)
ggo = enrichGO(entrez$ENTREZID, OrgDb = org.Hs.eg.db, ont = "ALL", readable = TRUE)
# write.table(gene_df, "gene_table.txt", sep = "\t", row.names = FALSE, quote = FALSE)
# write.table(entrez, "gene_entrezid.txt", sep = "\t", row.names = FALSE, quote = FALSE)
# write_json(ggo@geneSets, "go_geneset.json")
```

Download the result table:

<a href="https://raw.githubusercontent.com/zhuchcn/sr8278/master/gene_table_goids.txt" class="btn btn-primary">Download</a>

## peak regions

```{r message=FALSE, warning=FALSE}
peakAnno_tx = annotatePeak(tx_granges, TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db", level = "gene")
peakAnno_ve = annotatePeak(ve_granges, TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db", level = "gene")
data.frame(
    Feature = peakAnno_tx@annoStat$Feature,
    tx = peakAnno_tx@annoStat$Frequency * peakAnno_tx@peakNum,
    ve = peakAnno_ve@annoStat$Frequency * peakAnno_ve@peakNum
) %>%
    arrange(ve) %>%
    mutate(Feature = factor(Feature, levels = Feature)) %>%
    melt(id.vars = "Feature") %>%
    ggplot() +
    geom_col(aes(x = Feature, y = value, fill = variable), position = "dodge") +
    geom_text(aes(x = Feature, y = value + max(value)/15, 
                  label = formatC(value, format = "e", digit = 1), color = variable), 
              position = position_dodge(width = 1), size = 3) +
    coord_flip() +
    labs(y = "num peaks")
```



## RNAseq data

### 8278 48h

```{r}
rep1 = read.table("data/rsem_genes_results-8278_48h.txt", sep = "\t", header = TRUE)
rep2 = read.table("data/rsem_genes_results-8278_CON.txt", sep = "\t", header = TRUE)
fpkm_48h = data.frame(
    gene_id = rep1$gene_id,
    `48h` = rep1$FPKM,
    con = rep2$FPKM
) %>% filter(grepl("^ENSG", gene_id)) %>%
    filter(X48h != 0 & con != 0) %>%
    mutate(gene_id = gsub("\\.[0-9]+$", "", gene_id))
# down_reg_genes = fpmk$gene_id[fpmk$X48h < fpmk$con]
gene_48h = left_join(gene_df, fpkm_48h, by = "gene_id") %>% 
    filter(X48h < con & diff > 0) %>%
    mutate(fpkm_diff = con - X48h)
colnames(gene_48h) = c("gene_id", "gene_name", "npeaks_ve", "npeaks_tx", "npeaks_diff", "fpkm_48h", "fpkm_con", "fpkm_diff")
write.table(gene_48h, file = "gene_table-npeaks_fpmk-27ac_me1_vs_8278_48h.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```


Download the gene table with histone modification npeaks and RNAseq fpmk:

<a href="https://raw.githubusercontent.com/zhuchcn/sr8278/master/gene_table-npeaks_fpmk-27ac_me1_vs_8278_48h.txt" class="btn btn-primary">Download</a>

```{r}
gene_down = bitr(gene_48h$gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db, drop = TRUE)
gene_bg = bitr(gene_48h$gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db, drop = TRUE)
ego_48h = enrichGO(
    gene = gene_down$ENTREZID,
    OrgDb = org.Hs.eg.db,
    ont = "ALL",
    pvalueCutoff = 0.01,
    qvalueCutoff = 0.05,
    readable = TRUE
)
write.table(ego_48h@result, file = "go_enrichment-27ac_me1_vs_8278_48h.txt", sep = "\t", quote = FALSE, row.names = FALSE)
ego_48h@result %>% dplyr::select(-ID, -geneID) %>% 
    DT::datatable() %>%
    DT::formatSignif(columns = 5:7, digits = 3)
```

Download the go enrichment result of genes that may decrease in me1 and 27ac histone modification and may decrease in gene expression:

<a href="https://raw.githubusercontent.com/zhuchcn/sr8278/master/go_enrichment-27ac_me1_vs_8278_48h.txt" class="btn btn-primary">Download</a>

### 8278 24h

```{r}
rep1 = read.table("data/rsem_genes_results-8278_24h.txt", sep = "\t", header = TRUE)
rep2 = read.table("data/rsem_genes_results-8278_CON.txt", sep = "\t", header = TRUE)
fpkm_24h = data.frame(
    gene_id = rep1$gene_id,
    `24h` = rep1$FPKM,
    con = rep2$FPKM
) %>% filter(grepl("^ENSG", gene_id)) %>%
    filter(X24h != 0 & con != 0) %>%
    mutate(gene_id = gsub("\\.[0-9]+$", "", gene_id))
gene_24h = left_join(gene_df, fpkm_24h, by = "gene_id") %>% 
    filter(X24h < con & diff > 0) %>%
    mutate(fpkm_diff = con - X24h)
colnames(gene_24h) = c("gene_id", "gene_name", "npeaks_ve", "npeaks_tx", "npeaks_diff", "fpkm_24h", "fpkm_con", "fpkm_diff")
write.table(gene_24h, file = "gene_table-npeaks_fpmk-27ac_me1_vs_8278_24h.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```


Download the gene table with histone modification npeaks and RNAseq fpmk:

<a href="https://raw.githubusercontent.com/zhuchcn/sr8278/master/gene_table-npeaks_fpmk-27ac_me1_vs_8278_24h.txt" class="btn btn-primary">Download</a>

```{r}
gene_down = bitr(gene_24h$gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db, drop = TRUE)
gene_bg = bitr(gene_24h$gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db, drop = TRUE)
ego_24h = enrichGO(
    gene = gene_down$ENTREZID,
    OrgDb = org.Hs.eg.db,
    ont = "ALL",
    pvalueCutoff = 0.01,
    qvalueCutoff = 0.05,
    readable = TRUE
)
write.table(ego_24h@result, file = "go_enrichment-27ac_me1_vs_8278_24h.txt", sep = "\t", quote = FALSE, row.names = FALSE)
ego_24h@result %>% dplyr::select(-ID, -geneID) %>% 
    DT::datatable() %>%
    DT::formatSignif(columns = 5:7, digits = 3)
```

Download the go enrichment result of genes that may decrease in me1 and 27ac histone modification and may decrease in gene expression:

<a href="https://raw.githubusercontent.com/zhuchcn/sr8278/master/go_enrichment-27ac_me1_vs_8278_24h.txt" class="btn btn-primary">Download</a>

### sh rev (don't even know what that means)

```{r}
sh4 = read.table("data/rsem_genes_results-sh4_rev.txt", sep = "\t", header = TRUE)
sh5 = read.table("data/rsem_genes_results-sh5_rev.txt", sep = "\t", header = TRUE)
sicon = read.table("data/rsem_genes_results-sicon_rev.txt", sep = "\t", header = TRUE)
fpkm_sh = data.frame(
    gene_id = sh4$gene_id,
    sh4 = sh4$FPKM,
    sh5 = sh5$FPKM,
    sicon = sicon$FPKM
) %>% filter(grepl("^ENSG", gene_id)) %>%
    filter(sh4 != 0 & sh5 != 0 & sicon != 0) %>%
    mutate(gene_id = gsub("\\.[0-9]+$", "", gene_id))
gene_sh = left_join(gene_df, fpkm_sh, by = "gene_id") %>% 
    filter(sh4 < sicon & sh5 < sicon & diff > 0) %>%
    mutate(fpkm_diff_sh4 = sicon - sh4, fpkm_diff_sh5 = sicon - sh5)
colnames(gene_sh) = c("gene_id", "gene_name", "npeaks_ve", "npeaks_tx", "npeaks_diff", "fpkm_sh4", "fpkm_sh5", "fpkm_sicon", "fpkm_diff_sh4", "fpkm_diff_sh5")
write.table(gene_sh, file = "gene_table-npeaks_fpmk-27ac_me1_vs_sh_rev.txt", sep = "\t", quote = FALSE, row.names = FALSE)
```


Download the gene table with histone modification npeaks and RNAseq fpmk:

<a href="https://raw.githubusercontent.com/zhuchcn/sr8278/master/gene_table-npeaks_fpmk-27ac_me1_vs_sh_rev.txt" class="btn btn-primary">Download</a>

```{r}
gene_down = bitr(gene_sh$gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db, drop = TRUE)
gene_bg = bitr(gene_sh$gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db, drop = TRUE)
ego_sh = enrichGO(
    gene = gene_down$ENTREZID,
    OrgDb = org.Hs.eg.db,
    ont = "ALL",
    pvalueCutoff = 0.01,
    qvalueCutoff = 0.05,
    readable = TRUE
)
write.table(ego_sh@result, file = "go_enrichment-27ac_me1_vs_sh_rev.txt", sep = "\t", quote = FALSE, row.names = FALSE)
ego_sh@result %>% dplyr::select(-ID, -geneID) %>% 
    DT::datatable() %>%
    DT::formatSignif(columns = 5:7, digits = 3)
```

Download the go enrichment result of genes that may decrease in me1 and 27ac histone modification and may decrease in gene expression:

<a href="https://raw.githubusercontent.com/zhuchcn/sr8278/master/go_enrichment-27ac_me1_vs_sh_rev.txt" class="btn btn-primary">Download</a>

